### ISSUE TITLE
Pluggable storages

### ISSUE THREAD

**Issue by SanderMertens** [SanderMertens/flecs#680]
**Describe the problem you are trying to solve.**
There are scenarios in which using a table-based storage has usability and/or performance issues:
- when stable pointers are required
- when a component would introduce lots of fragmentation
- when it is expensive to move
- when there are a lot of components to move

**Describe the solution you'd like**
An interface that makes it possible to use custom storage plugins which can be configured per component.

**Comment by xentripetal** [comment]
Its been a couple years since this has seen activity, wanted to check if this was still under consideration?

**Comment by SanderMertens** [comment]
Yep definitely! I'm currently working on v4, which removes a lot of blockers for this work :)

**Comment by xentripetal** [comment]
Great to hear, thank you! 

I'm not very experienced with C but I will take a look and see if I can help contribute to this. Is there a design doc for how you see this being implemented?

**Comment by SanderMertens** [comment]
There isn't, but you can see the design in broad strokes here: https://github.com/SanderMertens/flecs/tree/storage_refactor

**Comment by xentripetal** [comment]
@addbounty $25

Would love to see this to allow native storage in dotnet. Putting an initial seed in to help drive it.

**Comment by addbounty** [comment]
[![banner](https://bountybot.dev/api/display/github/SanderMertens/flecs/680/banner?v=163debed-2b7f-4cbe-be98-514ad46a0ccd)](https://bountybot.dev/github/SanderMertens/flecs/issues/680)
[![button](https://bountybot.dev/api/display/github/SanderMertens/flecs/680/button?v=163debed-2b7f-4cbe-be98-514ad46a0ccd)](https://bountybot.dev/github/SanderMertens/flecs/issues/680)

Make a Draft PR **early** so others can see you are working on it! To automatically create one:

```sh
# Using npx (installed if you have NodeJS/npm)
npx bountybot start SanderMertens/flecs#680

# Or, use cargo (installed if you have Rust)
cargo install bounty
bounty start SanderMertens/flecs#680
```

When merged, you will receive the bounty!

**Comment by addbounty** [comment]
[![banner](https://bountybot.dev/api/display/github/SanderMertens/flecs/680/banner?v=257bccfd-5150-48b4-9b93-be785595f271)](https://bountybot.dev/github/SanderMertens/flecs/issues/680)
[![button](https://bountybot.dev/api/display/github/SanderMertens/flecs/680/button?v=257bccfd-5150-48b4-9b93-be785595f271)](https://bountybot.dev/github/SanderMertens/flecs/issues/680)

Make a Draft PR **early** so others can see you are working on it! To automatically create one:

```sh
# Using npx (installed if you have NodeJS/npm)
npx bountybot start SanderMertens/flecs#680

# Or, use cargo (installed if you have Rust)
cargo install bounty
bounty start SanderMertens/flecs#680
```

When merged, you will receive the bounty!

**Comment by addbounty** [comment]
[![banner](https://bountybot.dev/api/display/github/SanderMertens/flecs/680/banner?v=92dd1e35-fe40-4428-a6bc-18e7078f4f3c)](https://bountybot.dev/github/SanderMertens/flecs/issues/680)
[![button](https://bountybot.dev/api/display/github/SanderMertens/flecs/680/button?v=92dd1e35-fe40-4428-a6bc-18e7078f4f3c)](https://bountybot.dev/github/SanderMertens/flecs/issues/680)

Make a Draft PR **early** so others can see you are working on it! To automatically create one:

```sh
# Using npx (installed if you have NodeJS/npm)
npx bountybot start SanderMertens/flecs#680

# Or, use cargo (installed if you have Rust)
cargo install bounty
bounty start SanderMertens/flecs#680
```

When merged, you will receive the bounty!
